### PR TITLE
[RW-2484][risk=no] Create help tips sidebar component

### PR DESCRIPTION
**Pull request by dolbeew**
Created `help-sidebar` component, designed to match this [Invision screen](https://projects.invisionapp.com/share/ZURSZVNW4RE#/screens/356269832). Used in the following pages/components:
- Data homepage
- Data Set page
- Cohort Builder
- Cohort Review Participants page
- Cohort Review Participant Detail page
- Concept homepage
- Concept Set Details page
Screen Shot 2019-09-06 at 11 08 13 AM

Also moved `sidebar-content` component from `detail-page` in cohort review to new `help-sidebar` component.

### CODE REVIEW

File: ui/src/app/cohort-review/detail-page/detail-page.tsx [hunk 1]
```diff
@@ -137,101 +55,20 @@ export const DetailPage = withCurrentWorkspace()(
...
       return <React.Fragment>
-        {!(participant && annotations && annotationDefinitions) && <SpinnerOverlay />}
-        {participant && annotations && annotationDefinitions && <React.Fragment>
-          <div className={'detail-page ' + (sidebarOpen ? 'sidebar-open' : '')}
-               style={{...styles.detailPage, ...(sidebarOpen ? styles.detailPageOpen : {})}}>
-            <DetailHeader participant={participant}>
-            </DetailHeader>
-            <DetailTabs>
-            </DetailTabs>
+        {!participant ? <SpinnerOverlay /> : <React.Fragment>
```
[the remark this answers is not in the document]
Comment (dolbeew) [reply ui/src/app/cohort-review/detail-page/detail-page.tsx]: Nice idea, done.

File: ui/src/app/cohort-review/detail-page/detail-page.tsx [hunk 2]
```diff
@@ -137,101 +55,20 @@ export const DetailPage = withCurrentWorkspace()(
...
-            annotationDefinitions={annotationDefinitions}
-            setAnnotationDefinitions={this.setAnnotationDefinitions}>
-          </EditAnnotationDefinitionsModal>}
-          {creatingDefinition && <AddAnnotationDefinitionModal
-            annotationDefinitions={annotationDefinitions}
-            onCancel={this.closeCreateDefinitionModal}
-            onCreate={this.definitionCreated}>
-          </AddAnnotationDefinitionModal>}
+          <HelpSidebar location='reviewParticipantDetail' participant={participant}
+            setParticipant={this.setParticipant} />
```
[the remark this answers is not in the document]
Comment (dolbeew) [reply ui/src/app/cohort-review/detail-page/detail-page.tsx]: Yeah, I usually define the function as an arrow function, but I agree this is better.

File: ui/src/app/cohort-review/sidebar-content/sidebar-content.component.tsx [hunk 3]
```diff
@@ -351,11 +384,21 @@ export const SidebarContent = fp.flow(
...
               const filtered = fp.remove({cohortAnnotationDefinitionId}, annotations);
-              setAnnotations(filtered.concat(update.annotationId ? [update] : []));
+              this.setState({annotations: filtered.concat(update.annotationId ? [update] : [])});
             }
           }}
         />;
       })}
-    </div>;
+      {editingDefinitions && <EditAnnotationDefinitionsModal
+          onClose={this.closeEditDefinitionsModal}
```
[the remark this answers is not in the document]
Comment (dolbeew) [reply ui/src/app/cohort-review/sidebar-content/sidebar-content.component.tsx]: Done.

File: ui/src/app/cohort-review/sidebar-content/sidebar-content.component.tsx [hunk 4]
```diff
@@ -351,11 +384,21 @@ export const SidebarContent = fp.flow(
...
+          annotationDefinitions={annotationDefinitions}
+          setAnnotationDefinitions={(v) => this.setState({annotationDefinitions: v})}>
+      </EditAnnotationDefinitionsModal>}
+      {creatingDefinition && <AddAnnotationDefinitionModal
+          annotationDefinitions={annotationDefinitions}
+          onCancel={() => this.setState({creatingDefinition: false})}
+          onCreate={this.definitionCreated}>
```
[the remark this answers is not in the document]
Comment (dolbeew) [reply ui/src/app/cohort-review/sidebar-content/sidebar-content.component.tsx]: Done.

File: ui/src/app/cohort-search/modal/modal.component.css [hunk 5]
```diff
@@ -9,6 +9,7 @@
   visibility: hidden;
   transform: scale(1.1);
   transition: visibility 0.25s linear, opacity 0.25s 0s, transform 0.25s;
+  z-index: 102;
```
[the remark this answers is not in the document]
Comment (dolbeew) [reply ui/src/app/cohort-search/modal/modal.component.css]: The side bar was showing up in front of the overlay for this modal. I started at 100 for the sidebar content, then 101 for the icon bar on the right.

File: ui/src/app/components/help-sidebar.tsx [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+      sidebarOpen: false,
+      searchTerm: undefined
+    };
+  }
+
+  onIconClick = (icon: string) => {
+    let {activeIcon, sidebarOpen} = this.state;
+    sidebarOpen = !(icon === activeIcon && sidebarOpen);
+    if (sidebarOpen) {
+      this.setState({activeIcon: icon, sidebarOpen});
```
[the remark this answers is not in the document]
Comment (dolbeew) [reply ui/src/app/components/help-sidebar.tsx]: It's the same as `this.setState({activeIcon: icon, sidebarOpen: sidebarOpen});`

File: ui/src/app/components/help-sidebar.tsx [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+  constructor(props: Props) {
+    super(props);
+    this.state = {
+      activeIcon: undefined,
+      sidebarOpen: false,
+      searchTerm: undefined
+    };
+  }
+
+  onIconClick = (icon: string) => {
```
[the remark this answers is not in the document]
Comment (dolbeew) [reply ui/src/app/components/help-sidebar.tsx]: Done.

File: ui/src/app/components/help-sidebar.tsx [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    let {activeIcon, sidebarOpen} = this.state;
+    sidebarOpen = !(icon === activeIcon && sidebarOpen);
+    if (sidebarOpen) {
+      this.setState({activeIcon: icon, sidebarOpen});
+    } else {
+      this.hideSidebar();
+    }
+  }
+
+  hideSidebar = () => {
```
[the remark this answers is not in the document]
Comment (dolbeew) [reply ui/src/app/components/help-sidebar.tsx]: Done.

File: ui/src/app/components/help-sidebar.tsx [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+                    {content.content.map((item, i) =>
+                      <p key={i} style={styles.contentItem}>{item}</p>)
+                    }
+                  </div>;
+              })}
+            </div>)}
+          </div>
+          <div style={contentStyle('annotations')}>
+            {!!participant &&
+              <SidebarContent participant={participant} setParticipant={setParticipant} />
```
[the remark this answers is not in the document]
Comment (dolbeew) [reply ui/src/app/components/help-sidebar.tsx]: Done

File: ui/src/app/components/help-sidebar.tsx [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+      if (!sidebarOpen) {
+        this.setState({activeIcon: undefined});
+      }
+    }, 300);
+  }
+
+  render() {
+    const {location, participant, setParticipant} = this.props;
+    const {activeIcon, sidebarOpen} = this.state;
+    const sidebarContainerStyles = {...styles.sidebarContainer, zIndex: activeIcon ? 100 : -1};
```
[the remark this answers is not in the document]
Comment (dolbeew) [reply ui/src/app/components/help-sidebar.tsx]: I defined them here so I can change them based on `activeIcon` and `sidebarOpen` state properties. Should I just have separate properties in `styles` for each scenario? 

Something like

`
const styles = {
...
sidebarContainer: {...},
sidebarContainerActive: {...},
...
}
`

then in render 

`<div style={activeIcon ? styles.sidebarContainerActive : styles.sidebarContainer}>`

File: ui/src/app/components/help-sidebar.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (als364) [reply ui/src/app/components/help-sidebar.tsx]: Could you change the name of the `sidebarOpen` in this scope to something like `newSidebarOpen` and explicitly set `this.state.sidebarOpen` instead?

[the remark this answers is not in the document]
Comment (dolbeew) [reply ui/src/app/components/help-sidebar.tsx]: Yep, done.

File: ui/src/app/components/help-sidebar.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (dolbeew) [reply ui/src/app/components/help-sidebar.tsx]: Got it, done.

**Pull request merged**